src/vocabulary/processor.rs

```diff
@@ -0,0 +1,402 @@
+/// '¢' == '\u{00A2}' -> 0xA2 == 162
+/// '£' == '\u{00A3}' -> 0xA3 == 163
+/// ...
+/// 'ª' == '\u{00AA}' -> 0xAA == 170
+/// '«' == '\u{00AB}' -> 0xAB == 171
+/// '¬' == '\u{00AC}' -> 0xAC == 172
+/// ---
+/// 'Ń' == '\u{0143}' -> 0xAD == 173
+/// ---
+/// '®' == '\u{00AE}' -> 0xAE == 174
+/// '¯' == '\u{00AF}' -> 0xAF == 175
+/// '°' == '\u{00B0}' -> 0xB0 == 176
+/// ...
+/// 'ý' == '\u{00FD}' -> 0xFD == 253
+/// 'þ' == '\u{00FE}' -> 0xFE == 254
+/// 'ÿ' == '\u{00FF}' -> 0xFF == 255
+/// ---
+static CHAR_MAP: Lazy<HashMap<char, u8>> = Lazy::new(|| {
+    let mut char_map = HashMap::with_capacity(256);
+    let mut key = 0x100u32;
+    for byte in 0..=255u8 {
+        let char = byte as char;
+        if matches!(
+            char, '!'..='~' | '\u{00A1}'..='\u{00AC}' | '\u{00AE}'..='\u{00FF}',
+        ) {
+            char_map.insert(char, byte);
+        } else if let Some(ch) = char::from_u32(key) {
+            char_map.insert(ch, byte);
+            key += 1;
+        }
+    }
+    char_map
+});
+
+/// Token processor to adjust tokens according to the tokenizer's level.
+#[derive(Debug)]
+pub(crate) struct TokenProcessor {
+    level: TokenProcessorLevel,
+}
+
+/// Recognizes different tokenizer's levels.
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) enum TokenProcessorLevel {
+    /// Matches byte level tokenizer (e.g., gpt2).
+    Byte,
+    /// Matches byte fallback tokenizer (e.g., llama), which have <0x__> tokens for
+    /// all __ >= 0x80 to represent incomplete UTF-8 sequences.
+    ByteFallback(Mods),
+}
+
+/// Modifications to be applied by `TokenProcessor`of `ByteFallback` level.
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct Mods {
+    spacechar: char,
+}
+
+/// Default string modification to be applied by `TokenProcessor` of `ByteFallback` level.
+static DEFAULT_MODS: Mods = Mods { spacechar: ' ' };
+
+impl Mods {
+    /// Apply default modifications.
+    fn apply_default(&self, token: String) -> String {
+        let to = DEFAULT_MODS.spacechar.to_string();
+        token.replace(self.spacechar, &to)
+    }
+}
+
+/// Local structure to be deserialized into from HF's `ReplaceDecoder` in order to get a replace pattern.
+#[derive(Debug, Deserialize)]
+struct ReplaceDecoder {
+    content: String,
+    pattern: ReplacePattern,
+}
+
+impl ReplaceDecoder {
+    fn space_replacement(&self) -> Option<char> {
+        if self.content != " " {
+            return None;
+        }
+        match &self.pattern {
+            ReplacePattern::String(pattern) => {
+                let mut chars = pattern.chars();
+                let char = chars.next();
+                if let Some(replacement) = char {
+                    if chars.next().is_none() {
+                        return Some(replacement);
+                    }
+                }
+                None
+            }
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+enum ReplacePattern {
+    String(String),
+}
+
+impl TokenProcessor {
```

Comment:
Wdyt about moving this impl block right after the definition of `TokenProcessor`? It makes it easier to discover the impls (i.e., you jump to definition and all impls are just below the definition, compared to looking for other places for impls).